languages/tree-sitter-stack-graphs-java/src/stack-graphs.tsg

```diff
@@ -185,28 +359,95 @@ attribute node_symbol = node            => symbol = (source-text node), source_n
   edge member -> def
 }
 
+(method_declaration (formal_parameters (_) @param)) @method
+{
+  edge @param.lexical_scope -> @method.lexical_scope
+  edge @method.lexical_scope -> @param.def
+  edge @method.defs -> @param.def
+}
+
+(formal_parameter type: (_) @type (_) @id) @param
```

Comment:
I wouldn't have minded rolling formal parameters into the `method_declaration` rule as well, but they also occur in records and constructors, so it seemed worth splitting them out.

Reply:
There's an inherent tension between the top-down and bottom-up views of the tree here that I _mostly_ reconcile by favouring the bottom-up, but every so often it's worth the more complicated patterns.

Reply:
There's no (visible) single rule for types in the Java grammar, so the field name is used with a wildcard pattern. For `(_) @id`, I match with a wildcard not because there's no simple unambiguous rule, but as a way of future-proofing should the rule be renamed (which has happened recently in this grammar—tho not for something as ubiquitous as identifiers).